src/main/java/rx/internal/operators/OperatorMaterialize.java

```diff
@@ -29,41 +32,147 @@
             }
+        });
+        return parent;
+    }
 
-            @Override
-            public void onError(Throwable e) {
-                RxJavaPlugins.getInstance().getErrorHandler().handleError(e);
-                child.onNext(Notification.<T> createOnError(e));
-                child.onCompleted();
+    private static class ParentSubscriber<T> extends Subscriber<T> {
+
+        private final Subscriber<? super Notification<T>> child;
+
+        // guarded by this
+        private Notification<T> terminalNotification;
+        // guarded by this
+        private boolean busy = false;
+        // guarded by this
+        private boolean missed = false;
+
+        private volatile long requested;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<ParentSubscriber> REQUESTED = AtomicLongFieldUpdater
+                .newUpdater(ParentSubscriber.class, "requested");
+
+        ParentSubscriber(Subscriber<? super Notification<T>> child) {
+            this.child = child;
+        }
+
+        @Override
+        public void onStart() {
+            request(0);
+        }
+
+        void requestMore(long n) {
+            BackpressureUtils.getAndAddRequest(REQUESTED, this, n);
+            request(n);
+            drain();
+        }
+
+        @Override
+        public void onCompleted() {
+            synchronized (this) {
+                terminalNotification = Notification.createOnCompleted();
             }
+            drain();
+        }
 
-            @Override
-            public void onNext(T t) {
-                child.onNext(Notification.<T> createOnNext(t));
+        @Override
+        public void onError(Throwable e) {
+            synchronized (this) {
+                terminalNotification = Notification.createOnError(e);
             }
+            RxJavaPlugins.getInstance().getErrorHandler().handleError(e);
+            drain();
+        }
 
-        };
+        @Override
+        public void onNext(T t) {
+            child.onNext(Notification.createOnNext(t));
+            decrementRequested();
+        }
+
+        private void decrementRequested() {
+            // atomically decrement requested
+            while (true) {
+                long r = requested;
+                if (r == Long.MAX_VALUE) {
+                    // don't decrement if unlimited requested
+                    return;
+                } else if (REQUESTED.compareAndSet(this, r, r - 1)) {
+                    return;
+                }
+            }
+        }
+
+        private void drain() {
+            synchronized (this) {
+                if (busy) {
+                    // set flag to force extra loop if drain loop running
+                    missed = true;
+                    return;
+                } else {
+                    // is our turn to run drain loop so set missed to false
+                    // so we can detect missed drain calls
+                    missed = false;
```

Comment:
If not busy, missed is always false (due to exhaustive drain) so no need to set it here.